File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[][]}
     */
    levelOrder(root) {
        const result = []
        const dfs = (tree, level) => {
            if (tree == null) return
            
            if (result.length < level + 1) result.push([])
            result[level].push(tree.val)

            dfs(tree.left, level + 1)
            dfs(tree.right, level + 1)
        }

        dfs(root, 0)
        return result
    }
}

File: Data Structures & Algorithms/clone-graph/submission-0.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        cache = {}

        def dfs(node: Optional['Node']):
            if node is None:
                return node
            if node in cache:
                return cache[node]
            
            cloned = Node(node.val)
            cache[node] = cloned

            for n in node.neighbors:
                if cloned.neighbors is None:
                    cloned.neighbors = []
                cloned.neighbors.append(dfs(n))

            return cloned

        return dfs(node)
